src/framerate.rs

```diff
@@ -0,0 +1,265 @@
+use num::{FromPrimitive, ToPrimitive};
+use std::fmt;
+use std::fmt::Formatter;
+use std::str::FromStr;
+
+/// NTSC is the type of NTSC standard a framerate adheres to.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Ntsc {
+    /// False means this Framerate is not NTSC.
+    False(),
+    /// NonDrop means this Framerate is non-drop NTSC (no frame numbers are dropped to sync timecode
+    /// with real-world time - results in Timecode that drifts from true time).
+    NonDropFrame(),
+    /// DropFrame means this framerate is drop-frame NTSC (frames numbers are dropped periodically
+    /// to keep timecode in sync with real-world time).
+    DropFrame(),
+}
+
+impl Ntsc {
+    /// check returns whether this is any NTSC format (drop or non-drop).
+    pub fn check(self) -> bool {
+        self != Self::False()
+    }
+
+    /// is_non_drop returns true if this is an NTSC non-drop-frame specification.
+    pub fn is_non_dropframe(self) -> bool {
+        self == Self::NonDropFrame()
+    }
+
+    /// is_dropframe returns true if this is an NTSC drop-frame specification.
+    pub fn is_dropframe(self) -> bool {
+        self == Self::DropFrame()
+    }
+}
+
+/// Framerate is the rate at which a video file frames are played back.
+///
+/// Framerate is measured in frames-per-second (24/1 = 24 frames-per-second).
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Framerate {
+    value: num::Rational64,
+    ntsc: Ntsc,
+}
+
+impl Framerate {
+    /// playback is the rational representation of the real-world playback speed as a fraction
+    /// in frames-per-second
+    ///
+    /// ex: 24000/1001.
+    pub fn playback(&self) -> num::Rational64 {
+        self.value
+    }
+
+    /// timebase is the rational representation of the timecode timebase speed as a fraction in
+    /// frames-per-second.
+    ///
+    /// ex: 24/1.
+    pub fn timebase(&self) -> num::Rational64 {
+        // If this is an NTSC timebase, we need to round it to the nearest whole number.
+        if self.ntsc.check() {
+            return self.value.round();
+        }
+        self.value
+    }
+
+    /// ntsc is whether this is an NTSC-style time base (aka 23.98, 24000/1001, etc). It returns
+    /// an enum detailing if it is not NTSC or what type of NTSC flavor it is.
+    pub fn ntsc(&self) -> Ntsc {
+        self.ntsc
+    }
+
+    /// from_rational is the private method we are going to use for parsing from both timebases and
+    /// framerates.
+    fn from_rational(value: num::Rational64, ntsc: Ntsc, is_base: bool) -> Result<Self, String> {
+        validate_ntsc_value(value, ntsc, is_base)?;
+
+        let mut new_value = value;
```

Comment:
You could avoid a mut here but assigning new_value to a block {..}
```suggestion
        let new_value = match is_base && ntsc.check() {
            true => value.round() * 1000 / 1001,
            false => value
        };
```